### PR TITLE
Add Ara support

### PR DESCRIPTION
**Pull request by niwis**
Support Ara via a custom, parametrised accelerator interface.
- `cv64a6_imafdcv_sv39_config_pkg.sv` enables V extension
- Pre-processor constant `ARIANE_ACCELERATOR_PORT` enables the interface between CVA6 and Ara. This may be merged into X-interface at some point.
- FPU is bumped to a SIMD-compatible version

Backwards compatibility should be preserved. Once this is merged, we will change the reference of Ara upstream CVA6.

Merry Christmas!

Depends on #948, #1101, #1102, #1103, #1104, #1105

### CODE REVIEW

**Comment by MikeOpenHWGroup** [comment]
Thanks a _lot_ for this @niwis.  I am just seeing this now - it got lost in the noise of my travels to the RISC-V Summit and the holidays.

Question for @JeanRochCoulon - should this pull-request be targeted to the `dev` branch?

**Comment by JeanRochCoulon** [comment]
Great PR @niwis thank you very much.
Adding to the master branch is maybe too early due to the fact that the interface is not yet CV-X-IF. @mike, your idea to host this in a new dev branch is smart.
In the next week, @ASintzoff and me will be available to review the PR.
à bientôt !!

**Comment by JeanRochCoulon** [comment]
Hello @niwis 
The PR includes a lot of modifications. And the PR is not finished. That's why I have created a new cva6 branch called dev/ara_integration. Can you submit a PR to this new branch ?
See you

**Comment by niwis** [comment]
Thanks @JeanRochCoulon and @MikeOpenHWGroup! A few follow-ups from my side regarding this:
- Regarding CVXIF compatibility: This is a major modification and I am not aware of a timeline for this. This is why we discussed with mike (correct me if I misunderstood) to initially add the Ara *accelerator interface* next to CVXIF and then merge both in a second, independent step.
- The accelerator interface is fully parametrised. So, without defining `ARIANE_ACCELERATOR_PORT`, there would be no change to the current state or interfaces of CVA6.
- Merging this into a stale branch would not help with keeping the modifications in sync with upstream development. Moreover, we would lose flexibility in our development.

Let me know what you think. I would also be happy to discuss this further next week.
Best, Nils

**Comment by jennycoder-fly** [comment]
@niwis Hi, How can I use Ara in openpiton+ariane system？

File: core/commit_stage.sv
Comment (zarubaf) [inline core/commit_stage.sv:22]: Can you move everything from here to a separate commit module? `commit_stage_rvv` or something like this. That makes the change easier and separated. You can wire the `dirty_fp_state_o` on the modules port in the top.

File: core/controller.sv
Comment (zarubaf) [inline core/controller.sv:28]: Can we move the entire change to the same (new) module above and only provide an input `ait_acc_store_i` which we use as `wait_acc_store_q` in the `halt_o`?

File: core/csr_regfile.sv
Comment (zarubaf) [inline core/csr_regfile.sv:494]: @JeanRochCoulon  That's good and equivalent to the original code. It needs to be off.

File: core/csr_regfile.sv
Comment (zarubaf) [inline core/csr_regfile.sv:630]: This needs to be fully gated by an `ifdef` right? Otherwise, the CSR will still be present and not throw an illegal instruction exception.

File: core/issue_read_operands.sv
Comment (zarubaf) [inline core/issue_read_operands.sv:68]: Can we move most of the logic to a module outside? I have the feeling just `stall` is needed. We can have a `acc_stall_i`, `acc_ready_i`. This would make the change to the module less intrusive.

File: core/load_unit.sv
Comment (zarubaf) [inline core/load_unit.sv:22]: Well the same for all the other stall signals :-)

File: core/serdiv.sv
Comment (zarubaf) [inline core/serdiv.sv:167]: Could we make this default for both? Or is it locking up CVA6. Not sure I fully understand the comment.

**Review by zarubaf — CHANGES_REQUESTED** [review]
Thanks so much for this initial contribution. I've given this a first pass and reviewed particular with respect to separating concerns between CVA6 base and the Ara contribution, so that we have a path forward to merge this. 

Please, everybody, feel free to comment.

File: core/decoder.sv
Comment (jennycoder-fly) [reply core/decoder.sv:1213]: Thanks so much

File: core/axi_shim.sv
Comment (JeanRochCoulon) [inline core/axi_shim.sv:255]: As reported in #948 , the CVA6 AXI does not support the MODIFIABLE transaction

File: core/cva6.sv
Comment (JeanRochCoulon) [inline core/cva6.sv:310]: When possible, do not use ifdef commands. Take example on cvxif to avoid this command

File: core/serdiv.sv
Comment (JeanRochCoulon) [reply core/serdiv.sv:167]: serdiv is instantiated in mult module which does not set the STABLE_HANDSHAKE parameter. Can we remove the modification ?

**Review by JeanRochCoulon — CHANGES_REQUESTED** [review]

**Comment by niwis** [comment]
Thank you, @zarubaf and @JeanRochCoulon, for your initial review! As discussed in today's technical meeting, we await a decision on whether you want to merge this feature with the currently implemented accelerator port (next to CVXIF). If so, we will be happy to
- implement your feedback,
- rebase onto the latest version of the master branch (new conflicts have emerged since this PR was opened),
- and to ensure the current scalar regression tests (Github + Thales Gitlab) and Ara's CI pass.

Looking forward to hearing back!

**Comment by niwis** [comment]
As suggested, some previous modifications of this PR have been moved to #1101, #1102, #1103, #1104, and #1105. Once they are merged, we can rebase this PR and continue with the steps above https://github.com/openhwgroup/cva6/pull/1024#issuecomment-1448762948. Let us know if more changes in this PR should be treated separately.

Cheers,
Nils

**Comment by zarubaf** [comment]
Thanks @niwis, much appreciated!

File: core/ariane.sv
Comment (niwis) [reply core/ariane.sv]: Continued in #948

File: core/include/ariane_pkg.sv
Comment (niwis) [reply core/include/ariane_pkg.sv]: Continued in #1104

File: vendor/pulp-platform_fpnew.vendor.hjson
Comment (niwis) [reply vendor/pulp-platform_fpnew.vendor.hjson]: Continued in #1102

**Comment by JeanRochCoulon** [comment]
Very good @niwis. At 9pm, at the time we usually eat (or drink), the CI will run on all your PRs. Let's see the result on Tomorrow.

**Comment by niwis** [comment]
Let's see tomorrow, then. Bon appétit ! :-)

**Comment by zarubaf** [comment]
I think we should have all dependencies of this one in @niwis - can we do a final rebase so that we are clean with the diff?

Then we'll review again and hopefully be all a happy CVA6 family :-D

**Comment by domenicw** [comment]
Thanks for all of this work, this looks really great. Regarding our (@zarubaf, @michael-platzer, and me) proposal for the multiple accelerator interface problem, I see only minimal changes that need to happen in this PR.

**First**, as @zarubaf has already mentioned in his review, the `acc_dispatcher` module should be moved out of the `ex_stage` and into the top module (`cva6`). All special signals that would be needed from the `ex_stage` (which there seem to none at the moment) should rather be routed out of it and into the new top level `acc_dispatcher`.

**Second**, there should only be a single `cva6` extension interface. With this PR the accelerator interface is added to the ports of the top module. These should be removed again and instead the `cvxif_req_o` and `cvxif_resp_i` ports should be used. For that, the top module parameter list would need to gain two new parameters:

```SystemVerilog
parameter type cvxif_req_t  = cvxif_pkg::cvxif_req_t,
parameter type cvxif_resp_t = cvxif_pkg::cvxif_resp_t
```

And then in the ports list:

```SystemVerilog
output cvxif_req_t  cvxif_req_o,
input  cvxif_resp_t cvxif_resp_i
```

This way the default would still be the x-interface, but Ara could just replace the parameters and use its own signal types. This should not lead to any problems in this PR, since only the x-interface or the accelerator interface can be enabled at a time. Though we need to make sure that the `cvxif_req_o` signals are tied to zero if neither of the two interface are enabled.
The `cvxif_req_o` and `cvxif_resp_i` signals would then be directly routed to the `acc_dispatcher` module.

---

These changes should be enough for this PR. In a next step, the x-interface should be adapted to use this new accelerator interface regime. Meaning it should gain its own `cva6_accel_first_pass_decoder` module with everything currently residing in the decoder. Further, all of the x-interface logic currently in the `ex_stage` would move into a `acc_dispatcher` module. All other logic distributed in the different stages could then be merged with the `accelerator` logic of this PR. There is no need for these redundant signals, since all accelerator offloading interfaces mostly do the same things.

The `cva6` project is heavily pushing the x-interface. Therefore we propose that the default for the accelerator interface becomes the x-interface. Meaning the `cva6_accel_first_pass_decoder` and the `acc_dispatcher` implementations that live in this project will be the ones for the x-interface. If you want to use your custom interface (e.g. Ara accelerator interface), then these custom modules would live in your personal project. Bender could then gain a new target, which can be used to exclude the default accelerator interface of `cva6`. Something like:

```yaml
- target: not(exclude_default_xif)
  files:
    - core/cva6_accel_first_pass_decoder.sv
    - core/acc_dispatcher.sv
```

---

All of these changes should make it easier to do a custom accelerator interface implementation. The upstream `cva6` project would not needed to be touched in any way and no new ports would need to be added to the top level module. The only time the upstream repo would need to be touched is if new signals need to be routed into or out of the `acc_dispatcher`. But as we mentioned during our proposal, the port list these modules would be unstable. If a new port is added, it has to be correctly implemented for the x-interface (the upstream always has to work). But for everyone else with their custom implementation, updating to a newer `cva6` version might mean to also update their custom modules.

@niwis, @JeanRochCoulon, and everyone else, let me know what you think of these changes and if anything is unclear. If further explanation is needed, we could also do another face to face meeting.

---

Also, just as a small detail, we might want to revisit the naming of these two modules, the pre-decoder and the dispatcher. Just to make them more consistent.

**Comment by zarubaf** [comment]
@niwis do you maybe have an ETA for us on those changes? Would be super cool if we can get this is in 😄

**Comment by niwis** [comment]
I'll try to tackle this this weekend. Another rebase is done and Ara's CI is passing, so the only thing missing is implementing the feedback

File: core/commit_stage.sv
Comment (niwis) [reply core/commit_stage.sv]: done, moved to `acc_dispatcher`.

File: core/controller.sv
Comment (niwis) [reply core/controller.sv]: done, moved to `acc_dispatcher` and wired `halt` from there.

File: core/csr_regfile.sv
Comment (niwis) [reply core/csr_regfile.sv]: Right. I solved this by setting `update_access_exception=1'b` (the default behaviour) when this is disabled (https://github.com/openhwgroup/cva6/blob/646f33caee03a4e5a5f781b935a42fe9e7d959b2/core/csr_regfile.sv#LL829C29-L829C29). This avoids using `ifdef`s

File: core/csr_regfile.sv
Comment (niwis) [reply core/csr_regfile.sv]: done

File: core/ex_stage.sv
Comment (niwis) [reply core/ex_stage.sv]: That's indeed an artefact. I have removed it.

File: core/ex_stage.sv
Comment (niwis) [reply core/ex_stage.sv]: done

File: core/include/axi_intf.sv
Comment (niwis) [reply core/include/axi_intf.sv]: This is indeed not needed. I have dropped the change.

File: core/issue_read_operands.sv
Comment (niwis) [reply core/issue_read_operands.sv]: Done. moved to `acc_dispatcher` for now.

File: core/load_store_unit.sv
Comment (niwis) [reply core/load_store_unit.sv]: yes, dropped this diff

File: core/issue_stage.sv
Comment (niwis) [reply core/issue_stage.sv]: done. moved to `acc_dispatcher`

File: core/load_unit.sv
Comment (niwis) [reply core/load_unit.sv]: manual constant propagation :D

File: core/serdiv.sv
Comment (niwis) [reply core/serdiv.sv:167]: I don't think so. this is needed by Ara: https://github.com/pulp-platform/ara/blob/ee887bdb20ce396d438639a85bc2640718806fdc/hardware/src/lane/simd_div.sv#L313
perhaps @mp-17 can comment further

File: core/decoder.sv
Comment (niwis) [reply core/decoder.sv]: I moved as much as possible to `cva6_accel_first_pass_decoder`. I did not see a cleaner way to mux than in the updated proposal without changing half of the file (basically every direct assignment into the output struct). Other suggestions are welcome!

File: core/include/ariane_pkg.sv
Comment (niwis) [reply core/include/ariane_pkg.sv:170]: The problem is that we (still) need a precompiler constant for the ifdef in the port list. But perhaps we can remove it.

File: core/include/ariane_pkg.sv
Comment (niwis) [reply core/include/ariane_pkg.sv:184]: @mp-17 can you comment on this?

File: core/include/ariane_pkg.sv
Comment (niwis) [reply core/include/ariane_pkg.sv:299]: done, `core/include/acc_pkg.sv`.

**Comment by niwis** [comment]
Thank you @zarubaf @JeanRochCoulon @domenicw for your thorough review! I hope I was able to address most feedback. In particular:
1. Most new logic was moved into the new modules, foremost `acc_dispatcher`. This one is a bit bloated now, we can easily split it up or rename it if preferred.
2. The accelerator interface was removed and instead the cvxif is overloaded as suggested by @domenicw.

With 2., the `ifdef` is now only needed to add the invalidation interface. From my point of view, this interface could also be added unconditionally, which, however, would again change the interfaces. I will let the maintainers decide :-)

Let me know if there are further comments or suggestions!

File: core/cva6.sv
Comment (zarubaf) [inline core/cva6.sv:380]: Very nice!

File: core/include/ariane_pkg.sv
Comment (zarubaf) [reply core/include/ariane_pkg.sv:170]: I think if we can fold it into the structs, we can get rid of it completely.

File: corev_apu/src/ariane.sv
Comment (zarubaf) [inline corev_apu/src/ariane.sv:53]: Maybe it is time to deprecate this one? Can you use the `cva6` top for your use-case? Then we don't really need to change anything here, agree?

**Comment by zarubaf** [comment]
@niwis I am super, super happy with how this turned out. The implementation is really clean now. The last remaining thing is the invalidation interface, which IMHO we can fold into the structs as well. If I am mistaken I'd be still okay to go ahead and merge this. Thanks a lot for all the work you put into this!

@JeanRochCoulon can you help us with a bit more information on the failing Thales CI? I am sure it is only some setup thing.

File: corev_apu/src/ariane.sv
Comment (niwis) [reply corev_apu/src/ariane.sv]: Absolutely!

File: core/cva6.sv
Comment (niwis) [reply core/cva6.sv:46]: The only reason why they are separate now is that (at least in Ara's system) they don't interface with the accelerator (ara) itself but with the `axi_inval_filter`, which issues invalidation requests based on aw transactions and is located outside of Ara. It would be possible to split/pack the struct in the system, but that might be a bit misleading and confusing. Otherwise, we could also change this into a parametric `coherence_[req|resp]` struct that could be used for arbitrary coherence signals in the future. Those are just a few ideas, happy to implement the preferred option :-)

**Comment by JeanRochCoulon** [comment]
Sure, I will !

File: core/cva6.sv
Comment (zarubaf) [reply core/cva6.sv:46]: I definitely get the point, I think this needs a bit more thinking though and to not block this PR I'd prefer to keep it in the `cvxif_req_t` / `cvxif_resp_t` struct for the moment, then we do not add new ports that we potentially can't remove anymore. Does that make sense?

File: core/include/ariane_pkg.sv
Comment (mp-17) [reply core/include/ariane_pkg.sv:184]: Hello @JeanRochCoulon, 

From the specs: 
> Vector floating-point instructions require the presence of base scalar floating-point extensions corresponding to the supported vector floating-point element widths.

And, again, for the `V` extension in particular:
> The V extension requires the scalar processor implements the F and D extensions

So, we cannot completely disable CVA6's FPU if Ara is enabled.

They dropped the mandatory support for 16-bit floating-point in the last version of the specs, so we should modify this parameter assignment. Nevertheless, `XF16` should be asserted if Ara is deployed with `XF16` support.

File: core/serdiv.sv
Comment (mp-17) [reply core/serdiv.sv:167]: Yes, I modified it for Ara to keep the `in_rdy_o` asserted during the handshake and this is needed.
@zarubaf I think the stable handshake was creating a problem with CVA6 back then; I am not sure anymore. I launched a CI here with CVA6 using the stable handshake: https://github.com/pulp-platform/ara/actions/runs/5356243109

File: core/cva6.sv
Comment (niwis) [reply core/cva6.sv]: Makes sense! Done.

File: core/cva6.sv
Comment (niwis) [reply core/cva6.sv]: Done. The ifdef has been completely removed.

File: core/include/ariane_pkg.sv
Comment (niwis) [reply core/include/ariane_pkg.sv:170]: Done

File: core/include/ariane_pkg.sv
Comment (JeanRochCoulon) [reply core/include/ariane_pkg.sv:174]: From your explanation, I understand that ARA uses F and D extensions implemented inside CVA6. Understood.

**Comment by JeanRochCoulon** [comment]
Hello @niwis 
We started to review the code. Nice improvements ! Congratulations. 
When enabling the RVV, we faced a lack of files
Moreover, the overloaded cvxif package is not in GitHub PR

%Error: /home/jrcoulon/Projects/ara/core-v-verif/core-v-cores/cva6/core/decoder.sv:82:9: Cannot find file containing module: 'cva6_accel_first_pass_decoder'
   82 |         cva6_accel_first_pass_decoder i_accel_decoder (
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Cheers

**Comment by niwis** [comment]
Thank you @JeanRochCoulon for your review.

Regarding the missing file:
The accelerator first pass decoder is meant to be provided by the accelerator since it is specific to the accelerator/extension. For that reason, Ara's first pass decoder is located in the [Ara repo](https://github.com/pulp-platform/ara/blob/main/hardware/src/cva6_accel_first_pass_decoder.sv). I have added a stub to make the tools happy during elaboration. However, it should never be instantiated (the instantiation is guarded by `ENABLE_ACCELERATOR`). Let's see if this fixes the Thales CI.

Regarding the overloaded cvxif:
This PR parametrises the cvxif port type. The type needs to be overloaded when CVA6 is instantiated in a system that uses a different accelerator interface ([`ara_system`](https://github.com/pulp-platform/ara/blob/eeb5d5d4f31f5a3839d2c55ab87169164aec0e0b/hardware/src/ara_system.sv#L116-L117) in our case). What exactly is missing in this PR?
https://github.com/openhwgroup/cva6/blob/82e928952f2e416a7d9722a005ccad2311e01635/core/cva6.sv#L44-L45

**Comment by JeanRochCoulon** [comment]
@niwis The Thales CI was down from 2/3 days due to https://github.com/openhwgroup/cva6/pull/1273 PR. I reverted this PR to make VCS pass. Of course, the drawback on VCS should be investigated to fix Questa.
With this revert, the ARA CI should be better. Could you rebase this PR with master to include the revert to launch the CI please ?
Have a nice day !

**Comment by niwis** [comment]
Rebase is done @JeanRochCoulon

**Comment by JeanRochCoulon** [comment]
Hello @niwis 
accelerate_req_t and accelerate_resp_t are defined twice in the code: in acc_pkg and in ariane_pkg, I wonder whether it is ok ?

BTW, the UVM simulation is not working, I am investigating

**Comment by niwis** [comment]
Thanks, @JeanRochCoulon, for catching this. I have removed the old `accelerator_[req|resp]_t` definition in `ariane_pkg`. There may have been a type mismatch in `acc_dispatcher`, which should also be fixed now.

**Comment by JeanRochCoulon** [comment]
@niwis @zarubaf  The CSR compliance test are failed on master, so ARA is not guilty   :-)

On synthesis side, I will have a look

**Comment by zarubaf** [comment]
> @niwis @zarubaf The CSR compliance test are failed on master, so ARA is not guilty :-)
> 
> On synthesis side, I will have a look

**Comment by zarubaf** [comment]
> @niwis @zarubaf The CSR compliance test are failed on master, so ARA is not guilty :-)
> 
> On synthesis side, I will have a look

**Comment by zarubaf** [comment]
> @niwis @zarubaf The CSR compliance test are failed on master, so ARA is not guilty :-)
> 
> On synthesis side, I will have a look

I just wanted to say thank you and clicked everywhere - oops! 😅

**Comment by JeanRochCoulon** [comment]
For information, I posted a PR to pass the synthesis:
https://github.com/pulp-platform/cva6/pull/30

This should help.

**Comment by niwis** [comment]
Thanks a lot, @JeanRochCoulon! It is merged.

**Comment by zarubaf** [comment]
As far as I understood we are waiting for CI to re-trigger. Yesterday's CI was failing because of a full disk, which has been changed. Let's cross fingers for tonight :-D

**Comment by JeanRochCoulon** [comment]
We have fixed CI Today (I hope, to be confirmed this evening).

@niwis  A conflict must be resolved... What is the plan?  
1. To pass the CI before solving the conflict, wait for the CI fixed then solve the conflict or 
2. To solve it before this evening ?

**Comment by niwis** [comment]
It looks like the CI didn't trigger last night. @JeanRochCoulon I can rebase and resolve the conflicts, but it might introduce new regressions. Or we can try to get this version to pass first. As you prefer.

**Comment by JeanRochCoulon** [comment]
Ok, we can first pass the CI before rebasing.

PS: Thales CI is down, I try to fix it up Today

**Comment by JeanRochCoulon** [comment]
Good news, the Thales CI is green (all fails are understood and not related to ARA), you can rebase.

**Pull request merged**
